src/embedding/node2vec.jl

```diff
@@ -0,0 +1,159 @@
+const Alias = Tuple{SparseVector{Int}, SparseVector{Float64}}
+
+"""
+    node2vec(g; walks_per_node, len, p, q, dims)
+
+Computes node embeddings on graph `g`, as per [1]. Performs biased random walks on the graph,
+then computes word embeddings by treating those random walks like sentences.
+
+Returns an nv(g) x dims matrix of embeddings
+
+# Arguments
+
+- `g::FeaturedGraph`: The graph to perform random walk on.
+- `walks_per_node::Int`: Number of walks starting on each node,
+total number of walks is `nv(g) * walks_per_node`
+- `len::Int`: Length of random walks
+- `p::Real`: Return parameter from [1]
+- `q::Real`: In-out parameter from [1]
+- `dims::Int`: Number of vector dimensions
+
+
+[1] https://cs.stanford.edu/~jure/pubs/node2vec-kdd16.pdf
+"""
+function node2vec(g::FeaturedGraph; walks_per_node::Int=100, len::Int=5, p::Real=0.5, q::Real=0.5, dims::Int=128)
+    walks = simulate_walks(g; walks_per_node=walks_per_node, len=len, p=p, q=q)
+    model = walks2vec(walks,dims=dims)
+    vecs = []
+    println(typeof(model))
+    for i in 1:nv(g)
+        push!(vecs, get_vector(model, string(i)))
+    end
+    matrix = cat(vecs..., dims=2)
+    return matrix
+end
+
+"""
+Modified version of Node2Vec.learn_embeddings[1]. Uses
+a Julia interface[2] to the original word2vec C code[3].
+
+Treats each random walk like a sentence, and computed word
+embeddings using node ID as words.
+
+[1] https://github.com/ollin18/Node2Vec.jl
+[2] https://github.com/JuliaText/Word2Vec.jl
+[3] https://code.google.com/archive/p/word2vec/
+"""
+function walks2vec(walks::Vector{Vector{Int}}; dims::Int=100)
+    str_walks=map(x -> string.(x),walks)
+
+    if Sys.iswindows()
+        rpath = pwd()
+    else
+        rpath = "/tmp"
+    end
+    the_walks = joinpath(rpath,"str_walk.txt")
+    the_vecs = joinpath(rpath,"str_walk-vec.txt")
+
+    writedlm(the_walks,str_walks)
+    word2vec(the_walks,the_vecs,verbose=true,size=dims)
+    model=wordvectors(the_vecs)
+    rm(the_walks)
+    rm(the_vecs)
+    model
+end
```

Comment:
Should separate and move file I/O part to example. Try to keep source code clean.

Reply:
@yuehhua This is the interface for the Word2Vec library, it's a necessary component of every call of node2vec and can be easily hidden from the user, I don't think it should be included in the example. We have to write to a file because the Word2Vec library runs the official C code on a file of text.

Reply:
OK. I see it.